### PR TITLE
ensure stitching happens after eager deletion

### LINKED ISSUE #20030: 🐛 Bug Report: Catalog leaving stale relations after unregistering entities
### 📜 Description

When I create a **Component** and set the **spec.owner** as **Team A**. The[ EntityProcessors will generate the ownedBy and ownerOf relations](https://github.com/backstage/backstage/blob/5bf8bc9c1466837f810ec000255f46100937fff5/plugins/catalog-backend/src/modules/core/BuiltinKindsEntityProcessor.ts#L139), and the [Stitcher will make sure that those relations get pushed to the **entity.relations**](https://github.com/backstage/backstage/blob/5bf8bc9c1466837f810ec000255f46100937fff5/plugins/catalog-backend/src/stitching/Stitcher.ts#L186) in my **final_entities** object

When I **Delete** an entity, it will trigger the **[DefaultEntitiesCatalog.removeEntityByUid](https://github.com/backstage/backstage/blob/5bf8bc9c1466837f810ec000255f46100937fff5/plugins/catalog-backend/src/service/DefaultEntitiesCatalog.ts#L582C1-L609)**, and that will make sure to detect all of the related entities, and force those entities to be stitched together again, which will update the **relations** of the entity in the **final_entities** table.

However, when I **Unregister** the **Component**, it goes through the **[DefaultLocationStore.deleteLocation](https://github.com/backstage/backstage/blob/5bf8bc9c1466837f810ec000255f46100937fff5/plugins/catalog-backend/src/modules/core/DefaultLocationStore.ts#L89C1-L111C8)**, which in turn triggers the **[deleteWithEagerPruningOfChildren](https://github.com/backstage/backstage/blob/5bf8bc9c1466837f810ec000255f46100937fff5/plugins/catalog-backend/src/database/operations/provider/deleteWithEagerPruningOfChildren.ts#L28-L195)**, which removes all of the entities that would be orphaned from the database, including the relations.

However, the unregistration process does not triggers the **stitching** process again, which makes my **final_entities.final_entity.relations**  attribute to still point to entities that no longer exist in the catalog, causing a weird inconsistence.

The next time someone adds a new **Component** where the **spec.owner** is **Team A**, that will force stitching and it will end up fixing the stale relationship, but until that happens, the entity will contain stale relationships.

### 👍 Expected behavior

- When unregistering an entity that relates to other entities, the related entities need to be stitched together again, so their **relations** attribute contains up-to-date data.

### 👎 Actual Behavior with Screenshots

- When opening a Group page, the Group will show a warning about relationship to entities that no longer exist

image

### 👟 Reproduction steps

1. Create **Team A** as a Group
2. Create **Component B** as  Component
3. Have **Component B** have **Team A** as the owner
4. Unregister **Component B** from the Catalog
5. Validate that **component b** still shows up in the **relations** attribute of **Team A**

### 📃 Provide the context for the Bug.

I believe this PR https://github.com/backstage/backstage/pull/18062 could potentially set the foundation necessary to fix this issue.

### 🖥️ Your Environment

OS:   Linux 5.10.16.3-microsoft-standard-WSL2 - linux/x64
node: v18.17.0
yarn: 1.22.19
cli:  0.22.11 (installed)
backstage:  1.17.2

Dependencies:
  @backstage/app-defaults                               1.4.2
  @backstage/backend-app-api                            0.4.2, 0.5.1
  @backstage/backend-common                             0.18.4, 0.19.3
  @backstage/backend-dev-utils                          0.1.1
  @backstage/backend-openapi-utils                      0.0.3
  @backstage/backend-plugin-api                         0.5.1, 0.6.1
  @backstage/backend-tasks                              0.5.6
  @backstage/catalog-client                             1.4.3
  @backstage/catalog-model                              1.4.1
  @backstage/cli-common                                 0.1.12
  @backstage/cli-node                                   0.1.3
  @backstage/cli                                        0.22.11
  @backstage/config-loader                              1.4.0
  @backstage/config                                     1.0.8
  @backstage/core-app-api                               1.9.1
  @backstage/core-components                            0.12.5, 0.13.4
  @backstage/core-plugin-api                            1.5.3
  @backstage/dev-utils                                  1.0.19
  @backstage/errors                                     1.2.1
  @backstage/eslint-plugin                              0.1.3
  @backstage/integration-aws-node                       0.1.5
  @backstage/integration-react                          1.1.17
  @backstage/integration                                1.6.1
  @backstage/plugin-api-docs                            0.9.9
  @backstage/plugin-app-backend                         0.3.49
  @backstage/plugin-app-node                            0.1.1
  @backstage/plugin-auth-backend                        0.18.7
  @backstage/plugin-auth-node                           0.2.18
  @backstage/plugin-catalog-backend-module-github       0.3.4
  @backstage/plugin-catalog-backend                     1.12.1
  @backstage/plugin-catalog-common                      1.0.15
  @backstage/plugin-catalog-graph                       0.2.34
  @backstage/plugin-catalog-import                      0.9.12
  @backstage/plugin-catalog-node                        1.4.2
  @backstage/plugin-catalog-react                       1.8.2
  @backstage/plugin-catalog                             1.12.3
  @backstage/plugin-events-node                         0.2.10
  @backstage/plugin-github-actions                      0.6.3
  @backstage/plugin-github-pull-requests-board          0.1.16
  @backstage/plugin-home-react                          0.1.2
  @backstage/plugin-home                                0.5.6
  @backstage/plugin-kubernetes-backend                  0.11.4
  @backstage/plugin-kubernetes-common                   0.6.5
  @backstage/plugin-kubernetes                          0.10.1
  @backstage/plugin-org                                 0.6.12
  @backstage/plugin-permission-common                   0.7.7
  @backstage/plugin-permission-node                     0.7.12
  @backstage/plugin-permission-react                    0.4.14
  @backstage/plugin-proxy-backend                       0.3.1
  @backstage/plugin-scaffolder-backend                  1.16.2
  @backstage/plugin-scaffolder-common                   1.4.0
  @backstage/plugin-scaffolder-node                     0.2.1
  @backstage/plugin-scaffolder-react                    1.5.3
  @backstage/plugin-scaffolder                          1.14.3
  @backstage/plugin-search-backend-module-catalog       0.1.5
  @backstage/plugin-search-backend-module-techdocs      0.1.5
  @backstage/plugin-search-backend-node                 1.2.5
  @backstage/plugin-search-backend                      1.4.1
  @backstage/plugin-search-common                       1.2.5
  @backstage/plugin-search-react                        1.6.4
  @backstage/plugin-search                              1.3.5
  @backstage/plugin-tech-insights-backend-module-jsonfc 0.1.33
  @backstage/plugin-tech-insights-backend               0.5.15
  @backstage/plugin-tech-insights-common                0.2.11
  @backstage/plugin-tech-insights-node                  0.4.7
  @backstage/plugin-tech-insights                       0.3.14
  @backstage/plugin-tech-radar                          0.6.7
  @backstage/plugin-techdocs-backend                    1.6.6
  @backstage/plugin-techdocs-module-addons-contrib      1.0.17
  @backstage/plugin-techdocs-node                       1.7.5
  @backstage/plugin-techdocs-react                      1.1.9
  @backstage/plugin-techdocs                            1.6.7
  @backstage/plugin-user-settings                       0.7.7
  @backstage/release-manifests                          0.0.9
  @backstage/test-utils                                 1.4.2
  @backstage/theme                                      0.2.19, 0.4.1
  @backstage/types                                      1.1.0
  @backstage/version-bridge                             1.0.4

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by freben**
Fixes #20030 by marking all entities for stitching who had relations to something that'd be deleted by this operation. The root of the problem was that deletion of relations was working just fine (through database CASCADE), but `final_entities` wasn't being updated accordingly through stitching.

This looks like a large refactor, but it's mostly about breaking out a function that finds what to delete as a separate thing (that implementation is entirely unchanged), and then adding the new process-and-stitch markers as needed around those refs before deletion.

In summary:

- `findDescendantsThatWouldHaveBeenOrphanedByDeletion` is a verbatim unchanged copy of what used to be the inner nested query
- `markEntitiesAffectedByDeletionForStitching` is the actual new addition

### CODE REVIEW

**Comment by sonikro** [comment]
So, I checked out your branch locally, and linked the package to my backstage instance. I ran backstage in debug-mode and I was able to confirm that this code works. When unregistering components, its now not only cascade deleting everything but its also forcing stitching to happen in the related entities. With the code in this PR, the Bug I reported in https://github.com/backstage/backstage/issues/20030 is no longer happening.

I tested both Unregistering and Hard Delete, and both are now re-stitching the related entities.

Great job @freben !!!

I was thinking we'd have to use a reference to the Stitcher class to force stitching, but I didn't know we could update the **hash** to **force-stitching**.

**Comment by freben** [comment]
Thanks for verifying!

Yeah eventually I want to experiment with using db triggers for this so that entities are always marked for stitching as a side effect of other db changes - no matter what the original cause was. Some day!

**Review by alde — APPROVED** [review]

**Comment by sonikro** [comment]
Any way this can be released as a "next" version so we can start using it before the next official release ?

**Comment by freben** [comment]
Everything that gets merged always goes into the next releases actually. They are made off of master.
